control-plane/api-gateway/controllers/finalizer_test.go

```diff
@@ -0,0 +1,77 @@
+package controllers
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestEnsureFinalizer(t *testing.T) {
+	finalizer := "test-finalizer"
+
+	cases := map[string]struct {
+		initialFinalizers []string
+		finalizerToAdd    string
+		shouldUpdate      bool
+	}{
+		"should update":     {[]string{}, finalizer, true},
+		"should not update": {[]string{finalizer}, finalizer, false},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			// It doesn't matter what the object is, as long as it implements client.Object.
+			// A Pod was as good as any other object here.
+			testObj := &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "test-obj",
+					Finalizers: tc.initialFinalizers,
+				},
+			}
+
+			client := fake.NewClientBuilder().WithObjects(testObj).Build()
+
+			didUpdate, err := EnsureFinalizer(context.Background(), client, testObj, tc.finalizerToAdd)
+
+			require.NoError(t, err)
+			require.Equal(t, tc.shouldUpdate, didUpdate)
+		})
+	}
+}
+
+func TestRemoveFinalizer(t *testing.T) {
+	finalizer := "test-finalizer"
+
+	cases := map[string]struct {
+		initialFinalizers []string
+		finalizerToRemove string
+		shouldUpdate      bool
+	}{
+		"should update":     {[]string{finalizer}, finalizer, true},
+		"should not update": {[]string{}, finalizer, false},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			// It doesn't matter what the object is, as long as it implements client.Object.
+			// A Pod was as good as any other object here.
+			testObj := &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "test-obj",
+					Finalizers: tc.initialFinalizers,
+				},
+			}
+
+			client := fake.NewClientBuilder().WithObjects(testObj).Build()
+
+			didUpdate, err := RemoveFinalizer(context.Background(), client, testObj, tc.finalizerToRemove)
+
+			require.NoError(t, err)
+			require.Equal(t, tc.shouldUpdate, didUpdate)
```

Comment:
nit: what do you think of using something like `expectedUpdateValue` and `actualUpdateValue`? I got a little tripped up on the naming here

Reply:
I kind of like this naming because I thought it was clear. Can you help me understand what tripped you up about the naming? I wonder if there's something I'm not seeing.

Reply:
I think it's mainly the `shouldUpdate`, like to me that reads as an assertion (like "should this update") rather than an expectation in conjunction with the names for the subtests being really close to this threw me off a little bit

Reply:
I kind of split the difference